src/catalog/tests/encoded.rs

```diff
@@ -0,0 +1,321 @@
+                let arbitrary_data = $strategy
+                    .new_tree(&mut $runner)
+                    .expect("unable to create arbitrary data")
+                    .current();
+                // Skip over generated data where kind is None because they are not interesting or
+                // possible in production. Unfortunately any of the inner fields can still be None,
+                // which is also not possible in production.
+                // TODO(jkosh44) See if there's an arbitrary config that forces Some.
+                if arbitrary_data.kind.is_some() {
+                    let raw_data: StateUpdateKindRaw = arbitrary_data.into();
+                    Some(raw_data)
+                } else {
+                    None
+                }
+            }};
+        }
+
+        let mut runner = proptest::test_runner::TestRunner::deterministic();
+        let mut raw_datas = Vec::with_capacity(ENCODED_TEST_CASES);
+        while raw_datas.len() < ENCODED_TEST_CASES {
+            let raw_data = match self {
+                ProtoVersion::V42 => raw_data!(
+                    mz_catalog::durable::upgrade::objects_v42::StateUpdateKind::arbitrary(),
+                    runner
+                ),
+                ProtoVersion::V43 => raw_data!(
+                    mz_catalog::durable::upgrade::objects_v43::StateUpdateKind::arbitrary(),
+                    runner
+                ),
+                ProtoVersion::V44 => raw_data!(
+                    mz_catalog::durable::upgrade::objects_v44::StateUpdateKind::arbitrary(),
+                    runner
+                ),
+                ProtoVersion::V45 => raw_data!(
+                    mz_catalog::durable::upgrade::objects_v45::StateUpdateKind::arbitrary(),
+                    runner
+                ),
+            };
+            if let Some(raw_data) = raw_data {
+                raw_datas.push(raw_data);
+            }
+        }
+        raw_datas
+    }
+
+    /// Roundtrip [`StateUpdateKindRaw`] through a specific versioned protobuf `StateUpdateKind`.
+    fn roundtrip(&self, raw: StateUpdateKindRaw) -> StateUpdateKindRaw {
+        match self {
+            ProtoVersion::V42 => {
+                mz_catalog::durable::upgrade::objects_v42::StateUpdateKind::try_from(raw)
+                    .unwrap()
+                    .into()
+            }
+            ProtoVersion::V43 => {
+                mz_catalog::durable::upgrade::objects_v43::StateUpdateKind::try_from(raw)
+                    .unwrap()
+                    .into()
+            }
+            ProtoVersion::V44 => {
+                mz_catalog::durable::upgrade::objects_v44::StateUpdateKind::try_from(raw)
+                    .unwrap()
+                    .into()
+            }
+            ProtoVersion::V45 => {
+                mz_catalog::durable::upgrade::objects_v45::StateUpdateKind::try_from(raw)
+                    .unwrap()
+                    .into()
+            }
+        }
+    }
+}
```

Comment:
This is all feels a bit hacky and hard to maintain. So if anyone has any suggestions, them I'll all ears.

Reply:
Maybe something like:

```
#[derive(Arbitrary, Serialize, Deserialize)]
enum AllVersionsStateUpdateKind {
    V42(mz_catalog::durable::upgrade::objects_v42::StateUpdateKind),
    ....
}
```

And then instead of roundtripping `objects_vXX::StateUpdateKind` (binary -> SourceData -> StateUpdateKindRaw -> StateUpdateKind -> ...), you could roundtrip `AllVersionsStateUpdateKind` using `StateUpdateKindRaw::{to_serde,from_serde}`, which is a bit of a compromise, but I think would be less boilerplate.

If you added a `AllVersionsStateUpdateKind::variant(&self) -> &str` method that returned e.g. `objects_v42`, then you could generate them by randomly creating `AllVersionsStateUpdateKind`s until you got the ones you needed (a bit sus) and you could also assert that they are the right type when roundtripping them (seems nice)

Reply:
You're suggestion gave me an idea that involves some macro magic that significantly reduces the boilerplate without having to compromise the round trips. Now whenever a new version is added or removed only a single line needs to be changed.